L1Trigger/Phase2L1GT/plugins/L1GTProducer.cc

```diff
@@ -0,0 +1,452 @@
+        gttPromptJetToken_(consumes<TkJetWordCollection>(config.getParameter<edm::InputTag>("GTTPromptJets"))),
+        gttDisplacedJetToken_(consumes<TkJetWordCollection>(config.getParameter<edm::InputTag>("GTTDisplacedJets"))),
+        gttPrimaryVertexToken_(consumes<VertexWordCollection>(config.getParameter<edm::InputTag>("GTTPrimaryVert"))),
+        gmtSaPromptMuonToken_(consumes<SAMuonCollection>(config.getParameter<edm::InputTag>("GMTSaPromptMuons"))),
+        gmtSaDisplacedMuonToken_(consumes<SAMuonCollection>(config.getParameter<edm::InputTag>("GMTSaDisplacedMuons"))),
+        gmtTkMuonToken_(consumes<TrackerMuonCollection>(config.getParameter<edm::InputTag>("GMTTkMuons"))),
+        cl2JetToken_(consumes<PFJetCollection>(config.getParameter<edm::InputTag>("CL2Jets"))),
+        cl2PhotonToken_(consumes<TkEmCollection>(config.getParameter<edm::InputTag>("CL2Photons"))),
+        cl2ElectronToken_(consumes<TkElectronCollection>(config.getParameter<edm::InputTag>("CL2Electrons"))),
+        cl2TauToken_(consumes<PFTauCollection>(config.getParameter<edm::InputTag>("CL2Taus"))),
+        cl2EtSumToken_(consumes<std::vector<l1t::EtSum>>(config.getParameter<edm::InputTag>("CL2EtSum"))),
+        cl2HtSumToken_(consumes<std::vector<l1t::EtSum>>(config.getParameter<edm::InputTag>("CL2HtSum"))) {
+    produces<P2GTCandidateCollection>("GTTPromptJets");
+    produces<P2GTCandidateCollection>("GTTDisplacedJets");
+    produces<P2GTCandidateCollection>("GTTPrimaryVert");
+
+    produces<P2GTCandidateCollection>("GMTSaPromptMuons");
+    produces<P2GTCandidateCollection>("GMTSaDisplacedMuons");
+    produces<P2GTCandidateCollection>("GMTTkMuons");
+
+    produces<P2GTCandidateCollection>("CL2Jets");
+    produces<P2GTCandidateCollection>("CL2Photons");
+    produces<P2GTCandidateCollection>("CL2Electrons");
+    produces<P2GTCandidateCollection>("CL2Taus");
+    produces<P2GTCandidateCollection>("CL2EtSum");
+    produces<P2GTCandidateCollection>("CL2HtSum");
+  }
+
+  void L1GTProducer::fillDescriptions(edm::ConfigurationDescriptions &description) {
+    edm::ParameterSetDescription desc;
+
+    edm::ParameterSetDescription scalesDesc;
+    L1GTScales::fillDescriptions(scalesDesc);
+    desc.add<edm::ParameterSetDescription>("scales", scalesDesc);
+
+    desc.addOptional<edm::InputTag>("GTTPromptJets");
+    desc.addOptional<edm::InputTag>("GTTDisplacedJets");
+    desc.addOptional<edm::InputTag>("GTTPrimaryVert");
+
+    desc.addOptional<edm::InputTag>("GMTSaPromptMuons");
+    desc.addOptional<edm::InputTag>("GMTSaDisplacedMuons");
+    desc.addOptional<edm::InputTag>("GMTTkMuons");
+
+    desc.addOptional<edm::InputTag>("CL2Jets");
+    desc.addOptional<edm::InputTag>("CL2Photons");
+    desc.addOptional<edm::InputTag>("CL2Electrons");
+    desc.addOptional<edm::InputTag>("CL2Taus");
+    desc.addOptional<edm::InputTag>("CL2EtSum");
+    desc.addOptional<edm::InputTag>("CL2HtSum");
```

Comment:
The constructor reads all these parameters unconditionally, so they would be better expressed as required parameters.